pkg/dataplane/service.go

```diff
@@ -160,38 +162,60 @@ func EnsureServices(ctx context.Context, helper *helper.Helper, instance *datapl
 	return nil
 }
 
-// CheckGlobalServiceExecutionConsistency - Check that global services are defined only once in all nodesets, report and fail if there are duplicates
-func CheckGlobalServiceExecutionConsistency(ctx context.Context, helper *helper.Helper, nodesets []dataplanev1.OpenStackDataPlaneNodeSet) error {
+// Dedupe services to deploy
+// Multiple Services of same ServiceType/ServiceName in a nodeset
+// Global Services in multiple NodeSets for a deployment
+func DedupeServices(ctx context.Context, helper *helper.Helper,
+	nodesets []dataplanev1.OpenStackDataPlaneNodeSet,
+	serviceOverride []string) (map[string][]string, error) {
+	var nodeSetServiceMap = make(map[string][]string, 0)
 	var globalServices []string
-	var allServices []string
-
-	for _, nodeset := range nodesets {
-		allServices = append(allServices, nodeset.Spec.Services...)
-	}
-	for _, svc := range allServices {
-		service, err := GetService(ctx, helper, svc)
+	if len(serviceOverride) != 0 {
+		services, err := dedupe(ctx, helper, serviceOverride, globalServices)
 		if err != nil {
-			helper.GetLogger().Error(err, fmt.Sprintf("error getting service %s for consistency check", svc))
-			return err
+			return nil, err
 		}
-
-		if service.Spec.DeployOnAllNodeSets {
-			if serviceInList(service.Name, globalServices) {
-				return fmt.Errorf("global service %s defined multiple times", service.Name)
+		nodeSetServiceMap[AllNodeSets] = services
```

Comment:
This line is missing coverage in functional tests. Since it's such an important feature, I think it does deserve it.

Reply:
Well, there was not functional test for `serviceOverride` in deployment, so I did not consider it. Anyway, added one now to cover this.

Reply:
The code you have introduced substantially increases complexity of the scenario. We need to ensure everything behaves properly. Thanks for taking care of that.